script/package.json

```diff
@@ -10,10 +10,10 @@
     "coffeelint": "1.15.7",
     "colors": "1.1.2",
     "donna": "1.0.16",
-    "electron-chromedriver": "^6.0.0",
-    "electron-link": "^0.4.2",
-    "electron-mksnapshot": "^6.0.0",
-    "electron-packager": "^14.2.1",
+    "electron-chromedriver": "^9.0.0",
+    "electron-link": "0.4.1",
+    "electron-mksnapshot": "^9.0.2",
+    "electron-packager": "12.2.0",
```

Comment:
I'm not familiar with what these modules do, so I'm not sure if it's important. But this seems to downgrade :arrow_down: `electron-link` and `electron-packager`.

Reply:
You are most welcome 🙇‍♂️ I am actually trying to build Atom on CI and locally on Mac (Mac build passes but it's doesn't start, linux fails and windows build/test process seeing some undefined variables), so I merged master in case I missed up anything during resolving conflicts.

As for the above comment about the modules, which ones you are not familiar with `electron-link` and `electron-packager`?

Reply:
Yeah, I don't know `electron-link` or `electron-packager` well.

I read their release notes and I suppose any version should work.

- https://github.com/electron/electron-packager/releases
- https://github.com/atom/electron-link/compare/v0.4.1...v0.4.2

Maybe bump `electron-packager` to v14 again? 🤷

Reply:
Also, I think certain things need to be cleared out every build. I am in the habit of doing:
`rm -rf ./node_modules script/node_modules apm/node_modules` before every build. This might be unnecessary, but I do think `rm -rf script/node_modules` can help.

Reply:
`electron-packager` is used to create cross-platform executables based on an electron source code.

I haven't explored that much with `electron-link` but it helps create the snapshot to boost startup.

Although I don't think that either depends on the electron version, so I will try to bump them to latest.

Reply:
Exactly I was about to suggest running `script/clean`, that's what I normally do when building locally.

Reply:
I noticed that you bumped `electron-chromedriver` to 9, while this should match the target electron version, so I am going to try downgrading it to match the electron version, although your PR should do that already based on the electronVersion meta in package.json.

Reply:
I think my PR might have been flawed.

If `ELECTRON_CUSTOM_VERSION` is set (which it always is during `script/bootstrap`, the `postinstall` script doesn't run. So you can get stuck with an old version of `chromedriver` and `mksnapshot`.

Perhaps the `postinstall` script should run unconditionally/every time.

Reply:
Sounds great!

On the other hand (thinking out loud here), normally the electron version and it's dependencies should bet set properly anyway (having the logic to re-download binaries might be a bit more complex than we need).

I am thinking if my original PR would work with either of the below
- electron 6.0.0
- electron 6.1.12 + Setting ELECTRON_CUSTOM_VERSION

Reply:
I have successfully built the debian package and the Windows installer with my own branch: https://github.com/DeeDeeG/atom/commits/electron-6.1.12-merge-master

I'm not sure of the error here, but I think that branch should work.

Basically this way:
> - electron 6.1.12 + Setting ELECTRON_CUSTOM_VERSION

Reply:
@DeeDeeG maybe you could open a PR with your branch, we could test both PRs in parallel

Reply:
So the build on your branch is going smoothly however, oddly enough, it failed on Linux while continuing to run on Mac and Windows

https://github.visualstudio.com/Atom/_build/results?buildId=72328&view=logs&j=e2cf4b02-5697-54ad-cf7c-fc2a840d53af&t=b808d44d-d1c2-5399-c2e1-8c4b890e97af

Reply:
I noticed that failure.

```console
Packaging app for platform linux x64 using electron v6.1.12
Error: spawn unzip ENOENT
  at Process.ChildProcess._handle.onexit (internal/child_process.js:264:19)
  at onErrorNT (internal/child_process.js:456:16)
  at processTicksAndRejections (internal/process/task_queues.js:84:9)
```

Maybe just a flaky test? I am re-running the build on that branch locally and will see if I can reproduce.

I feel like "unzip something" should just work and not related to my changes. Might be trying to run that too early, when the thing it is meant to unzip isn't written to disk yet. Just a guess.

Reply:
Just got past that step locally:

```console
Packaging app for platform linux x64 using electron v6.1.12
Changing permissions for node files in /home/[user]/atom/out/atom-dev-1.50.0-dev-d5b207213-amd64
Copying non-ASAR resources to /home/[user]/atom/out/atom-dev-1.50.0-dev-d5b207213-amd64/resources
```

Reply:
Zipping can be super flaky, I remember running across this before (but let's hope it's just flaky), I will wait until the CI finishes though, mac artifacts is working 🎉 (still need to pass the tests but that's great progress)

Reply:
I can install the built `out/atom-amd64.deb`. It works, but as mentioned above, it is slow (busy GPU process). For me locally, this is working.

Reply:
It's good progress for sure! Thank you.

Regarding performance issues, are you able to shed more light on that?

- Performance comparison (Atom on electron 5 vs electron 6, other apps that are running on electron 6)
- Are you able to trace back the performance issue to a specific process?
- Are you able to reproduce this on any other OS or setup?

Reply:
Update about the linux error: this looks like a download error when downloading electron (flakiness)

Reply:
> Regarding performance issues, are you able to shed more light on that?

-  Performance comparison (Atom on electron 5 vs electron 6, other apps that are running on electron 6)
   - Atom on Electron 5: Very low idle CPU usage with Electron 5.
      - Even on `electron-6.1.12-merge-master` branch with just `electronVersion` changed to `5.0.13` in `package.json`. So only difference is the Electron 5. Low idle CPU with Electron 5. 
   - Atom on electron 6: One CPU core always at 100% if Atom is on screen. Idle CPU usage becomes low after a few moments when Atom is minimized.
   - VS Code on Electron 6: Low idle CPU usage.
      - Tested: [VS Code 1.40](https://code.visualstudio.com/updates/v1_40#_electron-60-update) (first version of VS Code with Electron 6) through [VS Code 1.43](https://code.visualstudio.com/updates/v1_43#_electron-70-update) (first version of VS Code with Electron 7).
- Are you able to trace back the performance issue to a specific process?
   - Yes, GPU process.
- Are you able to reproduce this on any other OS or setup?
  - No. Can only reproduce on my one machine where I have Linux installed. Old Atom CPU, no GPU.
  - Atom built from my branch runs well on Windows. Low idle CPU usage on Windows.

Reply:
Not sure if you saw my above comment about shedding light on the performance issue.

Reply:
@DeeDeeG thanks for heads-up, I didn't notice the comment.

I guess we can move forward and put a pin on the performance issue, let's see if we can reproduce this on other machines later on (I added it to the upgrade's checklist)

Reply:
I filed an issue you can pin if you like: https://github.com/atom/atom/issues/20924

Reply:
I made a PR for electron-link which bumps acorn. https://github.com/atom/electron-link/pull/26

The old electron-link fails to parse new JavaScript. This happens in #20965 too